sympy/core/basic.py

```diff
@@ -1641,6 +1641,38 @@ def rewrite(self, *args, **hints):
                 else:
                     return self
 
+    _constructor_postprocessor_mapping = {}
+
+    @classmethod
+    def _exec_constructor_postprocessors(cls, obj):
+        # WARNING: This API is experimental.
+
+        # This is an experimental API that introduces constructor
+        # postprosessors for SymPy Core elements. If an argument of a SymPy
+        # expression has a `_constructor_postprocessor_mapping` attribute, it will
+        # be interpreted as a dictionary containing lists of postprocessing
+        # functions for matching expression node names.
+
+        clsname = obj.__class__.__name__
+        postprocessors = defaultdict(list)
+        for i in obj.args:
+            try:
```

Comment:
@asmeurer had to put this `try` because some SymPy types are non-hashable (the Quantum gates contain mutable matrices in their arguments, I'm not going to fix that).